### PR TITLE
🚀Add SuperSimpleNet model

### PR DESCRIPTION
**Pull request by blaz-r**
## 📝 Description

- This PR adds the [SuperSimpleNet](https://arxiv.org/pdf/2408.03143) model implementation.
- I also verified the performance on MVTec AD and the results are very close tho the ones reported in the paper.

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

**Comment by samet-akcay** [comment]
This is amazing @blaz-r! Thanks a lot! I'll review it shortly

**Comment by blaz-r** [comment]
I have fixed the linter error and formated the files. The tests for SuperSimpleNet are still failing on onnx and openvino export. I'm not entirely sure how to fix this, but I'll take a lot when I get the time.

**Comment by blaz-r** [comment]
Tests are failing on imgaug, somethign with numpy version incompatiblity it seems.

**Comment by samet-akcay** [comment]
@blaz-r, we will need to remove `imgaug` dependency to bump the numpy version. I think this could be the scope of another PR. For now, it would be easier to stick to `numpy<2.0`

File: src/anomalib/models/image/supersimplenet/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+A ResNet-like feature extractor first extracts features, which are then upscaled and
+average-pooled to capture neighboring context. Features are further refined for anomaly detection task in the adaptor module.
+During training, synthetic anomalies are generated at the feature level by adding Gaussian noise to regions defined by the
+binary Perlin noise mask. The perturbed features are then fed into the segmentation-detection
+module, which produces the anomaly map and the anomaly score. During inference, anomaly generation is skipped, and the model
+directly predicts the anomaly map and score. The predicted anomaly map is upscaled to match the input image size
+and refined with a Gaussian filter.
+
+This implementation supports both unsupervised and supervised setting, but Anomalib currently supports only unsupervised learning.
```
Comment (samet-akcay) [inline src/anomalib/models/image/supersimplenet/README.md:21]: What are the missing points in Anomalib to support supervised setting

Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/README.md:21]: Right now I believe there are no standard supervised datasets. Another problem is the Folder dataset as it assumes that abnormal samples are always in test set:
https://github.com/openvinotoolkit/anomalib/blob/bcc0b439f616b13a8629cb64d8bf0f88fc9083a8/src/anomalib/data/image/folder.py#L174-L175

Another thing for full reproduction of SuperSimpleNet results is the fixed flipping augmentation and frequency sampling. This is however not necessary, but needed for best results. It's also not SuperSimpleNet specific, so might be worth considering if other supervised model will be supported.

Comment (samet-akcay) [reply src/anomalib/models/image/supersimplenet/README.md:21]: yeah, we would like to diversify the model pool, and include more learning types than one-class models. Thanks for the feedback. 
@abc-125, you might want to be aware of this discussion as you have recently worked on this stuff

Comment (abc-125) [reply src/anomalib/models/image/supersimplenet/README.md:21]: Thank you for adding me, it would be great to have supervised models and datasets in Anomalib. Recently, I looked at how to add a supervised [dataset](https://github.com/abc-125/vad), and it certainly would require changing some base structures, such as paths to folders (I guess we will need `abnormal_train_dir` and maybe renaming the rest to make it easier to understand, `normal_train_dir`, etc.):
https://github.com/openvinotoolkit/anomalib/blob/bcc0b439f616b13a8629cb64d8bf0f88fc9083a8/src/anomalib/data/image/folder.py#L195-L202

**Comment by blaz-r** [comment]
Sure. SuperSimpleNet used only `_rand_perlin_2d` function, which I think can now be replaced with the new `generate_perlin_noise`.

**Comment by samet-akcay** [comment]
@blaz-r can you pull the latest changes again? The tests should pass after that

**Comment by samet-akcay** [comment]
> Okay, I am currently at the ICPR. I will try to sort this out when I get some time.

ah nice, enjoy the conference!

**Comment by blaz-r** [comment]
This is now using the update perlin noise code and includes the latest v2 changes.

**Comment by blaz-r** [comment]
Can you run the checks for this please @djdameln @samet-akcay

**Comment by samet-akcay** [comment]
> Can you run the checks for this please @djdameln @samet-akcay

for some reason there is no option to trigger the ci :/

**Comment by samet-akcay** [comment]
> Can you run the checks for this please @djdameln @samet-akcay

@blaz-r, the tests seem to pass. I'll give another review and we could merge. Thanks for this great contribution!

**Review by samet-akcay — COMMENTED** [review]
I only have a single comment...

Anomalib's model naming convention is usually snake-case to camel case for model directory vs model class implementation. If this is followed, `supersimplenet` would be `super_simple_net` as the class name is `SuperSimpleNet`. Which one do you think is better in this case?

**Comment by blaz-r** [comment]
I think that `supersimplenet` would be better, since this is a single word. So I will rename the class to Supersimplenet.

**Comment by blaz-r** [comment]
I have now renamed the class to follow the naming convention.

**Review by samet-akcay — APPROVED** [review]

File: src/anomalib/models/image/supersimplenet/lightning_model.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                    "weight_decay": 0.00001,
+                },
+            ],
+        )
+        sched = MultiStepLR(
+            optim,
+            milestones=[int(self.trainer.max_epochs * 0.8), int(self.trainer.max_epochs * 0.9)],
+            gamma=0.4,
+        )
+        return [optim], [sched]
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/lightning_model.py:177]: We prefer to use full variables names,  i.e. `optimizer` and `scheduler`

File: src/anomalib/models/image/supersimplenet/lightning_model.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    @property
+    def learning_type(self) -> LearningType:
+        """Return the learning type of the model.
+
+        Returns:
+            LearningType: Learning type of the model.
+        """
+        return LearningType.ONE_CLASS
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/lightning_model.py:186]: I guess this would change when we add support for the supervised mode to the overall Anomalib pipeline. Maybe you could add a comment about this, as it might confuse users. Just a brief mention in the docstring would be sufficient.

File: src/anomalib/models/image/supersimplenet/torch_model.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+def init_weights(m: nn.Module) -> None:
+    """Init weight of the model.
+
+    Args:
+        m (nn.Module): torch module.
+    """
+    if isinstance(m, nn.Linear | nn.Conv2d):
+        nn.init.xavier_normal_(m.weight)
+    elif isinstance(m, nn.BatchNorm1d | nn.BatchNorm2d):
+        nn.init.constant_(m.weight, 1)
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/torch_model.py:139]: Again, I would prefer a full variable name here for readability.
```suggestion
def init_weights(module: nn.Module) -> None:
    """Init weight of the model.

    Args:
        module (nn.Module): torch module.
    """
    if isinstance(module, nn.Linear | nn.Conv2d):
        nn.init.xavier_normal_(module.weight)
    elif isinstance(module, nn.BatchNorm1d | nn.BatchNorm2d):
        nn.init.constant_(module.weight, 1)
```

File: src/anomalib/models/image/supersimplenet/torch_model.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        """
+        # dryrun
+        self.feature_extractor.eval()
+        with torch.no_grad():
+            features = self.feature_extractor(torch.rand(1, 3, 256, 256))
+        # sum channels
+        return sum(feature.shape[1] for feature in features.values())
+
+
+class FeatureAdaptor(nn.Module):
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/torch_model.py:210]: I would prefer `FeatureAdapter`, since we try to use American English throughout the code base. but I guess it would cause a discrepancy with the official implementation in your own repo. 

@samet-akcay @ashwinvaidya17 thoughts?

File: src/anomalib/models/image/supersimplenet/anomaly_generator.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+import torch.nn.functional as F  # noqa: N812
+from torch import nn
+
+from anomalib.data.utils.generators import generate_perlin_noise
+
+
+class SSNAnomalyGenerator(nn.Module):
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/anomaly_generator.py:13]: Not sure about the naming here. Why not simply use `AnomalyGenerator`? You don't use the `SSN` prefix for the other submodules.

File: src/anomalib/models/image/supersimplenet/torch_model.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    Args:
+        m (nn.Module): torch module.
+    """
+    if isinstance(m, nn.Linear | nn.Conv2d):
+        nn.init.xavier_normal_(m.weight)
+    elif isinstance(m, nn.BatchNorm1d | nn.BatchNorm2d):
+        nn.init.constant_(m.weight, 1)
+
+
+class FeatureExtractor(nn.Module):
```
Comment (djdameln) [inline src/anomalib/models/image/supersimplenet/torch_model.py:142]: You could consider cleaning up this file a bit by moving this class and `FeatureAdaptor` to a separate file (e.g. `features.py`). It's up to you, I don't have a strong preference.

**Review by djdameln — CHANGES_REQUESTED** [review]
Very nice and clean implementation, thanks a lot Blaž!

I only have some minor comments, mainly some naming issues.

File: src/anomalib/models/image/supersimplenet/torch_model.py (hunk 5 above)
Comment (samet-akcay) [reply src/anomalib/models/image/supersimplenet/torch_model.py:210]: as someone based in the UK, not sure what to say here 😁

File: src/anomalib/models/image/supersimplenet/lightning_model.py (hunk 2 above)
Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/lightning_model.py]: I update this and some other names as well.

File: src/anomalib/models/image/supersimplenet/torch_model.py (hunk 4 above)
Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/torch_model.py]: This is now changed.

File: src/anomalib/models/image/supersimplenet/torch_model.py (hunk 5 above)
Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/torch_model.py]: I changed this to FeatureAdapter, I think it's clear enough when compared with the official code 😄

File: src/anomalib/models/image/supersimplenet/anomaly_generator.py (hunk 6 above)
Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/anomaly_generator.py]: This is now just `AnomalyGenerator`.

File: src/anomalib/models/image/supersimplenet/torch_model.py (hunk 7 above)
Comment (blaz-r) [reply src/anomalib/models/image/supersimplenet/torch_model.py:142]: This is as is since the majority of the models currently follow this structure, so I adhered to it and I think it's best this way.

**Comment by blaz-r** [comment]
Thank you for the review. The comments are now addressed. Let me know if anything else needs to be sorted.

**Comment by blaz-r** [comment]
Seems like all checks pass but one that is failing due to some dead links. I think this is unrelated to my PR, or did I break something in docs?

**Comment by samet-akcay** [comment]
> Seems like all checks pass but one that is failing due to some dead links. I think this is unrelated to my PR, or did I break something in docs?

no worries, it is related to some new documentation URL, which is not on the readthedocs yet

**Review by djdameln — APPROVED** [review]
Thanks for the awesome work!
